### ISSUE TITLE
Include main.js in repo?

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#1584]
See https://github.com/graphhopper/graphhopper/pull/1572#issuecomment-477497841

**Comment by easbar** [comment]
This has probably been discussed a few times so far, but `main.js` seems to blow up the git repository size (because all its content is on the same line). This is the output I got running this nice [script](https://stackoverflow.com/a/42544963) showing git object sizes. I think the hashs are object ids not git commit hashs. Maybe it would be acceptable to remove main.js from the repository and include a manual build step before running the graphhopper server (could be included in `graphhopper.sh` as well). To be able to run without system-wide node/npm installation there is [nvm](https://github.com/creationix/nvm). Or maybe another easier way would be to not minify main.js (so not all content is on the same line and git can produce smaller diffs) for the repository and only use a minified version of main.js for production ?

```
0ffd88166777  1.3MiB core/files/N49E011.hgt.zip
53404479bc31  1.4MiB core/files/north-bayreuth.osm.gz
623adc98a9e2  1.4MiB core/files/N50E011.hgt.zip
8a76769f58ab  1.4MiB web/src/main/webapp/js/main.js
4dfa59baa221  1.4MiB reader-gtfs/files/beatty.osm
892c6aa9860e  2.0MiB reader-gtfs/files/rnv.zip
a2e129b7087a  2.3MiB web/src/main/resources/com/graphhopper/http/countries.geo.json
2c3edcd23bd2  2.3MiB core/files/spatialrules/countries.geo.json
ccfc0f941342  2.4MiB docs/slac-r-186-fixed-radius-near-neighbors.pdf
e6e61c15b914  2.8MiB web/src/main/webapp/js/main.js
2474cec2eb70  2.8MiB web/src/main/webapp/js/main.js
815c82541fc0  2.9MiB core/files/N48W125.hgt.zip
747c116e5f84  3.0MiB web/src/main/webapp/js/main.js
9b77c19142f3  3.0MiB web/src/main/webapp/js/main.js
7bea392542b3  3.0MiB web/src/main/webapp/js/main.js
491be81e73b2  3.1MiB reader-shp/src/test/resources/data/gis.osm_roads_free_1.shp
6b84456772ce  3.1MiB reader-shp/src/test/resources/data/malta-latest.osm.pbf
116ff9c6482f  3.5MiB reader-shp/src/test/resources/data/gis.osm_roads_free_1.dbf
2804fe44083f  3.6MiB pt-integration-tests/files/rnv.osm.pbf
b64626134bff  3.6MiB pt-integration-tests/files/rnv.zip
3a0a96c9703c  3.6MiB web/src/main/webapp/js/main.js
39cd164c65ff  3.7MiB web/src/main/webapp/js/main.js
c48da93cfe1a  4.0MiB web/src/main/resources/assets/js/main.js
65e62a49ea3f  4.0MiB web/src/main/resources/assets/js/main.js
```

**Comment by easbar** [comment]
From @karussell 

>   Maybe it would be acceptable to remove main.js from the repository

Yes. Let's remove it. Or maybe we still point to a released main.js and download it instead of building it.

**Comment by boldtrn** [comment]
If possible, I'd vote for keeping the main.js in the repository. This will make it easier for new users to experiment with GH. Keeping it in an unminified version will result in a significantly larger file. Most of the size is generated by external dependencies, their changes will result in large changes over time in the main.js file.

**Comment by easbar** [comment]
> This will make it easier for new users to experiment with GH

Yes I agree this is very important. But isn't there another way ? I think the download option sounds interesting especially since new users need to download an osm file anyway. Instead of downloading and saving it somewhere we could just include a hosted version via the `<script>` tag in index.html ?

**Comment by karussell** [comment]
> The drawback would of course be that the hosted main.js has to be kept up to date..

We could publish it to npm or something

**Comment by boldtrn** [comment]
I think we could build it on master on travis and push it to GithubPages or upload it to an object storage like Amazon S3.

**Comment by mprins** [comment]
You could also move it to [git LFS](https://help.github.com/en/articles/configuring-git-large-file-storage)

**Comment by michaz** [comment]
We could:
- Move it to its own repo 
- Build and deploy it with that toolchain
- Not commit the minified version
- Include the deployed version as a dependency in GraphHopper, for example as a WebJAR (no experience with that): https://www.webjars.org

**Comment by michaz** [comment]
> You could also move it to git LFS

Naaah. The ugly thing is that a compiled thing is committed to the repository at all. That it is too big is only a side-annoyance. ;-)

**Comment by boldtrn** [comment]
I think the biggest issue is that not everyone has a working npm installation and not everyone can/wants to have one. In order to build the main.js, npm is required to at least include the dependencies of the main.js.

> Move it (the web client) to its own repo

Personally I like this approach, this would be quite clean :). But everyone wo wants to quickly test GraphHopper would then need to checkout two different repos and start two different servers one for GraphHopper and one for the website. So that's not ideal either.

> Build and deploy it with its appropriate toolchain (npm probably?)

Yes and simply do this for every new commit to master on travis :). We could deploy this to a GitHub pages repo or maybe use something like jsDelivr.

**Comment by michaz** [comment]
> But everyone wo wants to quickly test GraphHopper would then need to checkout two different repos and start two different servers one for GraphHopper and one for the website.

No, they wouldn't! The current `.js` files are on the Java resource path. That means they can just as well be included as an (external) JAR, without making any difference to the user.

**Comment by karussell** [comment]
> We could deploy this to a GitHub pages repo

This would be indeed interesting.

> But everyone wo wants to quickly test GraphHopper

Why? If we have a separate repo for the demo (and client) we would highly likely publish this to npm (I'm unsure if we need or should publish JS via maven :)). 

With JS stuff on npm only two repos are necessary if the dev wants to customize the JS part too. 

The only problem I see in a separate repository is the increased maintenance for us developers (e.g. one change could require two commits etc). So I personally would prefer a separate folder in the root (in the same git repo) over a separate repo.

**Comment by easbar** [comment]
> Why? If we have a separate repo for the demo (and client) we would highly likely publish this to npm (I'm unsure if we need or should publish JS via maven :)).

Having it as a maven artifact like @michaz suggests would allow serving the JS from the graphhopper server, and all it takes is a maven include. also no extra server needed for local graphhopper maps. Publishing to npm requires npm installation which like @boldtrn maybe not everyone wants. 
I like [nvm](https://github.com/nvm-sh/nvm), its a shell script that downloads npm to a local folder without requiring root access or anything similar to a npm package and this could be set up in graphhopper.sh, but publishing via maven is still easier (even though it seems a bit odd to publish a maven module just for its JS contents).
Then again we could do it with neither maven nor npm and just put main.js anywhere, sth. like http://graphhopper.github.io/web/0.13-SNAPSHOT/main.js and 'download' it via a `<script>` tag in `index.html`.

> The only problem I see in a separate repository is the increased maintenance for us developers (e.g. one change could require two commits etc). So I personally would prefer a separate folder in the root (in the same git repo) over a separate repo.

Yes maybe keep the JS *sources* where they are at the moment, remove main.js and instead add a build step that creates main.js ?

**Comment by easbar** [comment]
webjar also sounds like what we want to do (but no idea)

>WebJars are client-side web libraries (e.g. jQuery & Bootstrap) packaged into JAR (Java Archive) files. 
    Explicitly and easily manage the client-side dependencies in JVM-based web applications

**Comment by michaz** [comment]
Meanwhile, I experimented with WebJars a bit. Preliminary result (subjective):
- Great if you produce simple free-handed JavaScript from your server, JSP-style, and you want to include a library or two, without using an extra JavaScript build at all.
- Not so great to refactor an existing JS project with its own build into.

My plan for the PT JavaScript client (not currently in the repo) is now to to what @easbar says and have its existing NPM build produce a JAR with the minified JS inside. Include that from `graphhopper`. Done. Will report. :-)

**Comment by karussell** [comment]
It is also worth noting that the dropwizard dependency to load resources from disk is not updated often: https://github.com/graphhopper/graphhopper/issues/1629

**Comment by easbar** [comment]
Fixed via #1641
